// @web3-react/walletlink-connector/dist/walletlink-connector.cjs.development.js
'use strict';

function _interopNamespace(e) {
  if (e && e.__esModule) { return e; } else {
    var n = {};
    if (e) {
      Object.keys(e).forEach(function (k) {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () {
            return e[k];
          }
        });
      });
    }
    n['default'] = e;
    return n;
  }
}

var abstractConnector = require('@web3-react/abstract-connector');

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

var CHAIN_ID = 1;
var WalletLinkConnector =
/*#__PURE__*/
function (_AbstractConnector) {
  _inheritsLoose(WalletLinkConnector, _AbstractConnector);

  function WalletLinkConnector(_ref) {
    var _this;

    var url = _ref.url,
        appName = _ref.appName,
        appLogoUrl = _ref.appLogoUrl;
    _this = _AbstractConnector.call(this, {
      supportedChainIds: [CHAIN_ID]
    }) || this;
    _this.url = url;
    _this.appName = appName;
    _this.appLogoUrl = appLogoUrl;
    _this.handleAccountsChanged = _this.handleAccountsChanged.bind(_assertThisInitialized(_this));
    return _this;
  }

  var _proto = WalletLinkConnector.prototype;

  _proto.handleAccountsChanged = function handleAccountsChanged(accounts) {
    {
      console.log("Handling 'accountsChanged' event with payload", accounts);
    }

    this.emitUpdate({
      account: accounts[0]
    });
  };

  _proto.activate = function activate() {
    try {
      var _temp3 = function _temp3() {
        _this3.provider.on('accountsChanged', _this3.handleAccountsChanged);

        return Promise.resolve(_this3.provider.send('eth_requestAccounts').then(function (accounts) {
          return accounts[0];
        })).then(function (account) {
          return {
            provider: _this3.provider,
            chainId: CHAIN_ID,
            account: account
          };
        });
      };

      var _this3 = this;

      var _temp4 = function () {
        if (!_this3.walletLink) {
          return Promise.resolve(new Promise(function (resolve) { resolve(_interopNamespace(require('walletlink'))); })).then(function (_ref2) {
            var WalletLink = _ref2["default"];
            _this3.walletLink = new WalletLink(_extends({
              appName: _this3.appName
            }, _this3.appLogoUrl ? {
              appLogoUrl: _this3.appLogoUrl
            } : {}));
            _this3.provider = _this3.walletLink.makeWeb3Provider(_this3.url, CHAIN_ID);
          });
        }
      }();

      return Promise.resolve(_temp4 && _temp4.then ? _temp4.then(_temp3) : _temp3(_temp4));
    } catch (e) {
      return Promise.reject(e);
    }
  };

  _proto.getProvider = function getProvider() {
    try {
      var _this5 = this;

      return Promise.resolve(_this5.provider);
    } catch (e) {
      return Promise.reject(e);
    }
  };

  _proto.getChainId = function getChainId() {
    return Promise.resolve(CHAIN_ID);
  };

  _proto.getAccount = function getAccount() {
    try {
      var _this7 = this;

      return Promise.resolve(_this7.provider.send('eth_accounts').then(function (accounts) {
        return accounts[0];
      }));
    } catch (e) {
      return Promise.reject(e);
    }
  };

  _proto.deactivate = function deactivate() {
    this.provider.removeListener('accountsChanged', this.handleAccountsChanged);
  };

  return WalletLinkConnector;
}(abstractConnector.AbstractConnector);

exports.WalletLinkConnector = WalletLinkConnector;
